File: src/ravel/matched_compute.py
"""Development-only matched-compute observations and fail-closed evaluation."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from .adaptation import ConstraintReport
from .policy import load_frozen_policy


Q20 = 1_048_576


@dataclass(frozen=True, slots=True)
class MatchedComputeObservation:
    candidate_training_evaluations: int
    matched_training_evaluations: int
    ratio_q20: int
    maximum_ratio_q20: int
    reference_available: bool
    threshold_identity: str
    comparator_identity: str
    partition_identity: str

    @classmethod
    def from_json(cls, value: Mapping[str, Any]) -> "MatchedComputeObservation":
        def count(name: str) -> int:
            result = value.get(name)
            if not isinstance(result, int) or result < 0:
                raise ValueError(f"matched-compute field {name} is malformed")
            return result

        ratio = count("ratio_q20")
        maximum = count("maximum_ratio_q20")
        available = value.get("reference_available")
        comparator = value.get("comparator_identity")
        partition = value.get("partition_identity")
        threshold = value.get("threshold_identity")
        if not isinstance(available, bool):
            raise ValueError("matched-compute availability is malformed")
        if not isinstance(comparator, str) or not comparator:
            raise ValueError("matched-compute comparator identity is malformed")
        if not isinstance(partition, str) or not partition:
            raise ValueError("matched-compute partition identity is malformed")
        if not isinstance(threshold, str) or not threshold:
            raise ValueError("matched-compute threshold identity is malformed")
        result = cls(
            count("candidate_training_evaluations"),
            count("matched_training_evaluations"),
            ratio,
            maximum,
            available,
            threshold,
            comparator,
            partition,
        )
        expected = (
            result.candidate_training_evaluations * Q20
            // result.matched_training_evaluations
            if result.matched_training_evaluations
            else 0
        )
        if result.ratio_q20 != expected:
            raise ValueError("matched-compute ratio does not reconstruct from raw counts")
        return result

    def evaluate(self) -> ConstraintReport:
        policy = load_frozen_policy()
        reasons: list[str] = []
        if self.threshold_identity != policy.threshold_identity:
            reasons.append("threshold_identity_mismatch")
        if not self.reference_available or self.matched_training_evaluations == 0:
            reasons.append("matched_compute_reference_unavailable")
        if self.maximum_ratio_q20 != policy.maximum_compute_ratio_q20:
            reasons.append("threshold_identity_mismatch")
        if self.ratio_q20 > policy.maximum_compute_ratio_q20:
            reasons.append("matched_compute_ratio")
        return ConstraintReport(not reasons, tuple(reasons))

File: crates/ravel-core/src/matched_compute.rs
//! Development-only matched-compute observations and fail-closed evaluation.

use crate::adaptation::ConstraintReport;
use crate::policy::{FrozenRavel06Policy, load_frozen_policy_from_root};
use ravel_contracts::Q20;
use ravel_contracts::reason::RejectionReason;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::path::Path;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum MatchedComputeError {
    #[error("{0}")]
    Invalid(String),
    #[error(transparent)]
    Policy(#[from] crate::policy::PolicyError),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MatchedComputeObservation {
    pub candidate_training_evaluations: i64,
    pub matched_training_evaluations: i64,
    pub ratio_q20: i64,
    pub maximum_ratio_q20: i64,
    pub reference_available: bool,
    pub threshold_identity: String,
    pub comparator_identity: String,
    pub partition_identity: String,
}

impl MatchedComputeObservation {
    pub fn from_value(value: &Value) -> Result<Self, MatchedComputeError> {
        let count = |name: &str| -> Result<i64, MatchedComputeError> {
            let result = value.get(name).and_then(Value::as_i64).ok_or_else(|| {
                MatchedComputeError::Invalid(format!("matched-compute field {name} is malformed"))
            })?;
            if result < 0 {
                return Err(MatchedComputeError::Invalid(format!(
                    "matched-compute field {name} is malformed"
                )));
            }
            Ok(result)
        };
        let require_str = |name: &str| -> Result<String, MatchedComputeError> {
            value
                .get(name)
                .and_then(Value::as_str)
                .filter(|item| !item.is_empty())
                .map(str::to_string)
                .ok_or_else(|| {
                    MatchedComputeError::Invalid(format!(
                        "matched-compute {name} identity is malformed"
                    ))
                })
        };
        let result = Self {
            candidate_training_evaluations: count("candidate_training_evaluations")?,
            matched_training_evaluations: count("matched_training_evaluations")?,
            ratio_q20: count("ratio_q20")?,
            maximum_ratio_q20: count("maximum_ratio_q20")?,
            reference_available: value
                .get("reference_available")
                .and_then(Value::as_bool)
                .ok_or_else(|| {
                    MatchedComputeError::Invalid("matched-compute availability is malformed".into())
                })?,
            threshold_identity: require_str("threshold_identity")?,
            comparator_identity: require_str("comparator_identity")?,
            partition_identity: require_str("partition_identity")?,
        };
        let expected = if result.matched_training_evaluations == 0 {
            0
        } else {
            result.candidate_training_evaluations * Q20 / result.matched_training_evaluations
        };
        if result.ratio_q20 != expected {
            return Err(MatchedComputeError::Invalid(
                "matched-compute ratio does not reconstruct from raw counts".into(),
            ));
        }
        Ok(result)
    }

    pub fn evaluate(
        &self,
        policy: &FrozenRavel06Policy,
    ) -> Result<ConstraintReport, MatchedComputeError> {
        let mut reasons = Vec::new();
        if self.threshold_identity != policy.threshold_identity()? {
            reasons.push(RejectionReason::THRESHOLD_IDENTITY_MISMATCH.into());
        }
        if !self.reference_available || self.matched_training_evaluations == 0 {
            reasons.push(RejectionReason::MATCHED_COMPUTE_REFERENCE_UNAVAILABLE.into());
        }
        if self.maximum_ratio_q20 != policy.maximum_compute_ratio_q20 {
            reasons.push(RejectionReason::THRESHOLD_IDENTITY_MISMATCH.into());
        }
        if self.ratio_q20 > policy.maximum_compute_ratio_q20 {
            reasons.push(RejectionReason::MATCHED_COMPUTE_RATIO.into());
        }
        Ok(ConstraintReport {
            passed: reasons.is_empty(),
            rejection_reasons: reasons,
        })
    }

    pub fn evaluate_from_root(&self, root: &Path) -> Result<ConstraintReport, MatchedComputeError> {
        let policy = load_frozen_policy_from_root(root)?;
        self.evaluate(&policy)
    }
}
